### ISSUE TITLE
How to 'activate' a BarChart Bar Tooltip or section programatically

### ISSUE THREAD

**Issue by RafaPolit** [recharts/recharts#784]
This is more of a question than a bug or feature request.

I am rendering a custom legend for a bar chart.  I want that, when I hover the custom legend items, to trigger the Tooltip Rendering and the background area highlight, just as if I had hovered with the mouse.

I have access to the legend index as expected, but how to trigger the same 'mouse enter' chain of events programmatically, as if I had just used the cursor?

Here's my case:
![screenshot from 2017-07-14 12-59-54](https://user-images.githubusercontent.com/4309019/28224446-ba0e14d8-6894-11e7-9e55-311b17dc9d3b.png)

Thanks in advanced,
Rafa.

**Comment by Saspect-IO** [comment]
The documentation already has your answer. 
Please see link below. Hope this helps,  if not,   I will test and put a sample code later if you are not able to implement it.
http://recharts.org/#/en-US/examples/LegendEffectOpacity

**Comment by RafaPolit** [comment]
No, this does not answer my question, I want to activate the Tooltip and 'active section' (the grey bar behind the bar in a barchcart) simulating the mouse over event ON THE BAR!!! Not on the legend.

The fact that I am using a legend is irrelevant. Consider it a button then. I need to make a button that triggers the same events as hovering over a bar on a bar chart.

That example holds nothing resembling my question.

**Comment by Saspect-IO** [comment]
My apologies for misinterpreting your question. I was just trying to help bro. 
I was reading the src code for the tooltips component . It has a set of default props.  With its active state set as false. Just a suggestion, maybe you could create a toggle button with a Boolean value that returns the tooltip component with a active state of true when the buttons selected state is true. And visa versa. 

https://github.com/recharts/recharts/blob/master/src/component/Tooltip.js
const defaultProps = {
  active: false,
  offset: 10,
  viewBox: { x1: 0, x2: 0, y1: 0, y2: 0 },
  coordinate: { x: 0, y: 0 },
  cursorStyle: {},
  separator: ' : ',
  wrapperStyle: {},
  itemStyle: {},
  labelStyle: {},
  cursor: true,
  isAnimationActive: !isSsr(),
  animationEasing: 'ease',
  animationDuration: 400,
  itemSorter: () => -1,
};

**Comment by RafaPolit** [comment]
I find it odd that no one has this requirement with the legend. How are you guys solving the need to display data upon hovering over the legend?

Thanks.

**Comment by abhayjeetgupta** [comment]
@RafaPolit I do have the same requirement. If u have solved it please help me.

**Comment by Bill-Q** [comment]
Same here. Would like to activate or select the data entry in the chart programmatically liking using a mouse to move over. Anyone has a solution for that?

**Comment by alfonsoperez** [comment]
@RafaPolit Did you ever manage to do this?

**Comment by Mathijs003** [comment]
+1

**Comment by edellucien** [comment]
I would be interested as well

**Comment by ckifer** [comment]
Taking a stab at this

**Comment by neomehta** [comment]
Is anybody working on this ?
I was looking something with a toggle feature on Legend click for BarChart, and It does not seem to be anything which satisfies it currently.

**Comment by ckifer** [comment]
@neomehta I tried and failed miserably. The codebase is confusing imo. My team is just changing the color of the bar for now... 

I can't find a way from the little digging I did to enable the 'cursor' programmatically. I just might need to spend more time in the code for it to make sense, but I don't have time to really :/

**Comment by neomehta** [comment]
@ckifer Same here, I have been trying it for almost a month now. however, I don't think I am any closer to finish it.

**Comment by ckifer** [comment]
@neomehta I'll let you know here if I make any progress but no promises. I'd totally be willing to work together on it or at least talk about it if you'd want to.

**Comment by Mathijs003** [comment]
One way I fixed it for now was to add a custom label to ReferenceLine like so : 

`<ReferenceLine x={} stroke={color} strokeDasharray="5 5" label={<CustomLabel/>}/>`

And then you can define svg elements to display text etc. I also used ForeignObject to add a white background: 
With x and y you can position your label.
```
const CustomLabel = (props: any) => {
    return (
      <>
        <foreignObject x={x - 15} y={y - 20} width="130" height={stacked ? '90' : '70'}>
          <div className="custom-graph-tooltip"/>
        </foreignObject>
        <text
          x={x}
          y={y}
        >
          {your text here}
        </text>
      </>
    )
  };
```

Not sure if it covers all your cases, but it did with mine.

**Comment by ckifer** [comment]
@Mathijs003 thanks for the example, that seems perfectly reasonable. I wish there was a way to display the original tooltip though and have the 'cursor' behind it highlight without having to hack a custom overlay. 

Adding some sort of flag or function to display the tooltip/highlight a certain index would be what I really need to try and add/get added.

I might make another issue to add an `onMouseEnter` and `onMouseLeave` event to the area of each bar that gets highlighted on hover (right now you can only access the entire graph or each individual bar).

**Comment by ckotyan** [comment]
I had a similar requirement (the tooltip cursor to be shown programatically). I ended up using the ReferenceArea construct.
<ReferenceArea x1={x1ColumnDataValue} x2={x2ColumnDataValue} />
This does not show the tooltip itself though.
My requirement was to show selectedBar with tooltip cursor like background.

**Comment by ElGoorf** [comment]
Same question. I was wondering if maybe they was a way to hack in via a `syncId`.

**Comment by Mredul-Hasan** [comment]
> This is more of a question than a bug or feature request.
> 
> I am rendering a custom legend for a bar chart. I want that, when I hover the custom legend items, to trigger the Tooltip Rendering and the background area highlight, just as if I had hovered with the mouse.
> 
> I have access to the legend index as expected, but how to trigger the same 'mouse enter' chain of events programmatically, as if I had just used the cursor?
> 
> Here's my case:
> ![screenshot from 2017-07-14 12-59-54](https://user-images.githubusercontent.com/4309019/28224446-ba0e14d8-6894-11e7-9e55-311b17dc9d3b.png)
> 
> Thanks in advanced,
> Rafa.

I need to display Legends name with color, just like your image. If you can enlighten me

**Comment by IanIsFluent** [comment]
> I have the same requirement. It would be great if we could pass `activeTooltipIndex` or `activeTooltipLabel` as a prop to the `<BarChart/>` element.
> 
> My use case is that I want to keep one section of a bar graph highlighted, with its cursor and tooltip active, after a user clicks on it. i.e. "freeze" the tooltip in place on one section of a graph, showing the cursor, even after the user has moved their mouse away to other sections. I can access the `activeTooltipIndex` and `activeTooltipLabel` properties from the `onClick` function on `<BarChart/>`, so it would be trivial to then pass them to the chart as props after a click event.

This might also help with testing tooltips! :)

**Comment by avshyz-gs** [comment]
Is there any update on this?

**Comment by kyle-marshall** [comment]
I was able to achieve a similar effect in a simple way after struggling with more hacky solutions.

The solution:

```tsx
<Tooltip defaultIndex={highlightIndex} />
```

To clear the highlighting, manually update the chart state using a ref:

```tsx
  useEffect(() => {
    if (!chartRef.current) {
      return;
    }
    if (highlightIndex === -1) {
      chartRef.current.setState(() => ({
        isTooltipActive: false,
      }));
      return;
    }
  }, [highlightIndex]);
  ```

Backstory: 

I was trying to simulate the mouse enter behavior, and was only getting partial success. Then I noticed the (undocumented?) `defaultIndex` prop of the Tooltip component might be useful:
https://github.com/recharts/recharts/blob/147ee259dcf609276fa6dcdce9ffb91431d1529c/src/chart/generateCategoricalChart.tsx#L1078

So I tried using this prop to control the active tooltip/highlight dynamically, and to my surprise it just worked.

**Comment by ckifer** [comment]
defaultIndex is a new-ish prop. It is documented in the storybook

**Comment by giltheodor** [comment]
Hello, Do you have any estimation when the defaultIndex in Tooltip feature will be available?

**Comment by ckifer** [comment]
@giltheodor the `defaultIndex` prop exists in the latest recharts, however it does not do as one expects as it only controls the first interaction (hence default). You can see the shortcomings of this [here](https://codesandbox.io/p/devbox/g5fhyp?migrateFrom=ks8k4g). This prop is still useful for rendering a tooltip index on initial render and then surrendering control over to the user.

**Comment by kyle-marshall** [comment]
> @giltheodor the `defaultIndex` prop exists in the latest recharts, however it does not do as one expects as it only controls the first interaction (hence default). You can see the shortcomings of this [here](https://codesandbox.io/p/devbox/g5fhyp?migrateFrom=ks8k4g). This prop is still useful for rendering a tooltip index on initial render and then surrendering control over to the user.

For me, the chart does indeed react whenever I change defaultIndex after the first time I set it -- is this a bug in the implementation then? The name seemed strange considering the behavior I get, I was hoping this was the intended behavior and it was just unfortunately named that instead of 'activeIndex'.

**Comment by ckifer** [comment]
@kyle-marshall I'm not sure what behavior you are getting, but the behavior of the linked sandbox is the intended behavior. This was added for accessibility reasons - please read PR https://github.com/recharts/recharts/pull/4070

**Comment by ckifer** [comment]
@julianna-langston correct me if I'm wrong, but your intention in the above PR was to allow programmatic control with `defaultIndex`? Or not? In retrospect the current functionality is confusing as shown in this conversation

**Comment by kyle-marshall** [comment]
> @kyle-marshall I'm not sure what behavior you are getting, but the behavior of the linked sandbox is the intended behavior. This was added for accessibility reasons - please read PR #4070

The sandbox is trying to set the same index each time.
If you modify `updateIndex` as follows you get the behavior where it does indeed continue to respond to changes after setting it the 1st time:

```typescript
  const updateIndex = () => {
    setIndex(prev => (prev + 1) % data.length);
  };
  ```
[updated sandbox](https://codesandbox.io/p/devbox/g5fhyp?migrateFrom=ks8k4g&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cly3sv0ei0008356jbui80mop%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cly3sv0eh0002356j5comz1e6%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cly3sv0eh0005356jkwb8aa9c%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cly3sv0eh0007356j0jgmogw1%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cly3sv0eh0002356j5comz1e6%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cly3sv0eh0001356j787d83d3%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cly3svslc003d356j7gkksxye%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A88%252C%2522startColumn%2522%253A9%252C%2522endLineNumber%2522%253A88%252C%2522endColumn%2522%253A18%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cly3sv0eh0002356j5comz1e6%2522%252C%2522activeTabId%2522%253A%2522cly3svslc003d356j7gkksxye%2522%257D%252C%2522cly3sv0eh0007356j0jgmogw1%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cly3sv0eh0006356j0isjhq96%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522Development%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cly3sv0eh0007356j0jgmogw1%2522%252C%2522activeTabId%2522%253A%2522cly3sv0eh0006356j0isjhq96%2522%257D%252C%2522cly3sv0eh0005356jkwb8aa9c%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cly3sv0eh0003356jhfgvwzdj%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Development%2522%257D%252C%257B%2522id%2522%253A%2522cly3sv0eh0004356jaebbx1x4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cly3l2xtx000adaf87jcm1b50%2522%257D%255D%252C%2522id%2522%253A%2522cly3sv0eh0005356jkwb8aa9c%2522%252C%2522activeTabId%2522%253A%2522cly3sv0eh0003356jhfgvwzdj%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by ckifer** [comment]
ah got it, thanks for the example. I always had kept trying a static value. so yep, still a limitation there but not as bad as I thought

**Comment by julianna-langston** [comment]
You should be able to pass in a prop to `defaultIndex` and have it just work. If that doesn't work, there might be a bug.

So,
```jsx
<Tooltip active defaultIndex={myIndex} />
```
should update with something like this:
```js
setMyIndex(myIndex+1)
```

**Comment by julianna-langston** [comment]
> For me, the chart does indeed react whenever I change defaultIndex after the first time I set it -- is this a bug in the implementation then? The name seemed strange considering the behavior I get, I was hoping this was the intended behavior and it was just unfortunately named that instead of 'activeIndex'.

It's called `defaultIndex`, because that's the default index *for the user*. Users are still able to move the tooltip around, but `defaultIndex` indicates where the tooltip should start. I didn't want to call it `activeIndex`, since that felt like it'd lock the tooltip in place.

**Comment by kyle-marshall** [comment]
> > For me, the chart does indeed react whenever I change defaultIndex after the first time I set it -- is this a bug in the implementation then? The name seemed strange considering the behavior I get, I was hoping this was the intended behavior and it was just unfortunately named that instead of 'activeIndex'.
> 
> It's called `defaultIndex`, because that's the default index _for the user_. Users are still able to move the tooltip around, but `defaultIndex` indicates where the tooltip should start. I didn't want to call it `activeIndex`, since that felt like it'd lock the tooltip in place.

Yes this name makes more sense now, I did not realize the limitation at first.

I think a possible alternate solution could be to have both an `activeIndex` on the tool tip and an `onMouseOverIndex` prop. If `activeIndex` is not defined, behavior remains uncontrolled (e.g. internally updating the index when mouse moves across chart), however if `activeIndex` is provided it becomes fully controlled by user and they would have to provide `onMouseOverIndex` callback if they care about responding to that interaction. This is more burden on the user if they truly only care about setting the default, but would be flexible enough for more use cases. Just some thoughts - I am abusing `defaultIndex` currently in a work project so I may be interested in trying to implement this.

**Comment by ckifer** [comment]
onMouseMove/onMouseEnter already exist so no need for new events I don't think. 

It's more complicated than it seems like it should be once you get into the code, but feel free to contribute. It has to make sense for everyone though.

**Comment by kyle-marshall** [comment]
> onMouseMove/onMouseEnter already exist so no need for new events I don't think.
> 
> It's more complicated than it seems like it should be once you get into the code, but feel free to contribute. It has to make sense for everyone though.

Nice, with the mouse event already existing, one 'activeIndex' prop allowing full control by parent component should technically be able to cover both use cases, but I agree it is much easier said than done, and I can live with the current behavior so it's not worth it to me yet to dive in. Anyways, thanks for all the replies!

**Comment by julianna-langston** [comment]
If you’re interested in contributing, note that we have an internal-only prop called `activeIndex`, so your change should be making that prop public (and then handling test cases and logic for different, related props). Still harder than it sounds, if you’ve never looked at the Recharts source code before.

**Comment by nhaebigkerber** [comment]
Has there been any progress on this issue? In my org, we are using mapbox to present a map of location data that is connected to a set of chart data. would love for users to be able to move through the map and update the corresponding chart tooltip.

**Comment by PavelVanecek** [comment]
@nhaebigkerber you can either set a `defaultIndex` prop on Tooltip which does the same thing as hovering over the bar with a mouse, or pass a `shape` prop to Bar and render anything you wish.

**Comment by nhaebigkerber** [comment]
@PavelVanecek would there be any possibility of a future API change allowing consumers to access the eventCenter directly? Before I saw your comment I had worked something out based on this [old solution](https://stackoverflow.com/questions/52855858/how-to-sync-recharts-chart-with-something-custom-that-is-not-a-recharts-graph) from stack overflow that calls `eventCenter.emit` with the index from the map point. Actually ended up being pretty simple but I did have to reach into the package's internals. The tooltip approach seems 'safer' for sure but also seems like a workaround to the underlying issue of not being able to access the synced tooltip's position directly.

**Comment by PavelVanecek** [comment]
If possible I would prefer to keep the declarative interface we have now, and drive the chart state that way instead of exposing eventEmitter. `defaultIndex` is not the most meaningful name (wasn't mine) but it is the way to control Tooltip programatically. Can you please elaborate on what feels workaround about it? It's quite straightforward from my point of view, pass the index, that's where Tooltip appears, job done.

**Comment by PavelVanecek** [comment]
Also for the record I want to state that importing the eventEmitter and listening on the events is not "totally safe" and can break without announcements. The event shape is not a public API. If you do that then I recommend to lock to specific patch version of Recharts (but then you're giving up on new features and bug fixes).

**Comment by nhaebigkerber** [comment]
yep definitely understand that it's not public and not safe. We are on v2 due to peer dependency problems but I will try the index suggestion when I get some time to work through that and upgrade to v3.

**Comment by nhaebigkerber** [comment]
I guess my rationale was that it would make sense to control the value directly via the internal event, but if the defaultIndex approach does the same thing and the sync value isn't used anywhere else then I guess there's no reason to add another public entry point.

**Comment by PavelVanecek** [comment]
Give it a try and see how it works. I have some ideas on how to improve this situation that don't involve exposing the events, so if you need some adjustments we can try that.
